File: back/src/main/java/com/back/domain/trend/score/TrendZScoreCalculator.java
package com.back.domain.trend.score;

import com.back.domain.trend.score.dto.WordFrequencyStatsDto;
import org.springframework.stereotype.Component;

@Component
public class TrendZScoreCalculator {

    public double calculate(WordFrequencyStatsDto baseline, WordFrequencyStatsDto current) {
        if (baseline.totalMessages() <= 0 || current.totalMessages() <= 0) {
            return 0.0;
        }
        double baselineRatio = calculateBaselineRatio(baseline);
        double currentRatio = calculateCurrentRatio(current);
        double pooledRatio = calculatePooledRatio(baseline, current);
        double standardError = calculateStandardError(pooledRatio, baseline, current);
        double zScore = (currentRatio - baselineRatio) / standardError;
        if (Double.isNaN(zScore) || Double.isInfinite(zScore)) {
            return 0.0;
        }
        return zScore;
    }

    private double calculateBaselineRatio(WordFrequencyStatsDto baseline) {
        return (double) baseline.frequency() / baseline.totalMessages();
    }

    private double calculateCurrentRatio(WordFrequencyStatsDto current) {
        return (double) current.frequency() / current.totalMessages();
    }

    private double calculatePooledRatio(WordFrequencyStatsDto baseline, WordFrequencyStatsDto current) {
        return (double) (baseline.frequency() + current.frequency()) / (baseline.totalMessages() + current.totalMessages());
    }

    private double calculateStandardError(double pooledRatio, WordFrequencyStatsDto baseline, WordFrequencyStatsDto current) {
        return Math.sqrt(pooledRatio * (1 - pooledRatio) * (1 / (double) current.totalMessages() + 1 / (double) baseline.totalMessages()));
    }
}

File: back/src/main/kotlin/com/back/domain/trend/score/TrendZScoreCalculator.kt
package com.back.domain.trend.score

import com.back.domain.trend.score.dto.WordFrequencyStatsDto
import org.springframework.stereotype.Component
import kotlin.math.sqrt

@Component
class TrendZScoreCalculator {
    fun calculate(baseline: WordFrequencyStatsDto, current: WordFrequencyStatsDto): Double {
        if (baseline.totalMessages <= 0 || current.totalMessages <= 0) {
            return 0.0
        }
        val baselineRatio = calculateBaselineRatio(baseline)
        val currentRatio = calculateCurrentRatio(current)
        val pooledRatio = calculatePooledRatio(baseline, current)
        val standardError = calculateStandardError(pooledRatio, baseline, current)
        val zScore = (currentRatio - baselineRatio) / standardError
        if (zScore.isNaN() || zScore.isInfinite()) {
            return 0.0
        }
        return zScore
    }

    private fun calculateBaselineRatio(baseline: WordFrequencyStatsDto): Double =
        baseline.frequency.toDouble() / baseline.totalMessages

    private fun calculateCurrentRatio(current: WordFrequencyStatsDto): Double =
        current.frequency.toDouble() / current.totalMessages

    private fun calculatePooledRatio(baseline: WordFrequencyStatsDto, current: WordFrequencyStatsDto): Double =
        (baseline.frequency + current.frequency).toDouble() / (baseline.totalMessages + current.totalMessages)

    private fun calculateStandardError(
        pooledRatio: Double,
        baseline: WordFrequencyStatsDto,
        current: WordFrequencyStatsDto
    ): Double =
        sqrt(pooledRatio * (1 - pooledRatio) * (1.0 / current.totalMessages + 1.0 / baseline.totalMessages))
}
